### PR TITLE
Adding Fee Calculation Tests

### PR DESCRIPTION
**Pull request by idky137**
Adds tests for fee calculation.

### CODE REVIEW

File: zingo-testutils/src/lib.rs [hunk 1]
```diff
@@ -243,6 +244,182 @@ pub async fn load_wallet(
...
 }
 
+#[derive(Debug)]
+pub struct TxActions {
+    pub transparent_tx_actions: usize,
+    pub sapling_tx_actions: usize,
+    pub orchard_tx_actions: usize,
+}
+
+/// Returnes number of notes used as inputs for txid as TxActions (transparent_notes, sapling_notes, orchard_notes).
```
Comment (Oscar-Pepper) [inline zingo-testutils/src/lib.rs]: returns typo

**Comment by idky137** [comment]
do not merge yet. removing reliance on do_list_txsummaries before ready to merge as this function is bugged!

File: zingo-testutils/src/lib.rs [hunk 2]
```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
     )
 }
 
+#[derive(Debug)]
+pub struct TxNotes {
```
Comment (fluidvanadium) [inline zingo-testutils/src/lib.rs]: maybe this should be called TxNotesCount?

**Review by fluidvanadium — COMMENTED** [review]
i dont know why the tests are running as if using the zip317 feature.

File: zingo-testutils/src/lib.rs (hunk 2 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs]: change made in [27144bd]

**Comment by idky137** [comment]
Ive removed the feature gate for fee calculation and commented out zip317 expected_fee, it seams that CI runs tests with all features enabled so these tests wont pass. 

When zip317 lands these tests will fail. The current expected_fee will need to be removed and zip317 expected_fee will need to be un-commented, but no new code needs to be written. Ive added comments to show this.

Should now be ready to merge.

change made in [8719fad]

File: zingo-testutils/src/lib.rs [hunk 3]
```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
...
+    }
+
+    TxNotesCount {
+        transparent_tx_notes: transparent_notes,
+        sapling_tx_notes: sapling_notes,
+        orchard_tx_notes: orchard_notes,
+    }
+}
+
+/// Returnes number of notes created in txid as TxActions (transparent_notes, sapling_notes, orchard_notes).
```
Comment (Oscar-Pepper) [inline zingo-testutils/src/lib.rs]: returns typo

File: zingo-testutils/src/lib.rs [hunk 4]
```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
...
 
+#[derive(Debug)]
+pub struct TxNotesCount {
+    pub transparent_tx_notes: usize,
+    pub sapling_tx_notes: usize,
+    pub orchard_tx_notes: usize,
+}
+
+#[derive(Debug)]
+pub struct TxActions {
```
Comment (Oscar-Pepper) [inline zingo-testutils/src/lib.rs]: TxActionsCount for consistency?

File: zingo-testutils/src/lib.rs [hunk 5]
```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
...
+/// Returns the total transfer value of txid.
+pub async fn total_tx_value(client: &LightClient, txid: &str) -> u64 {
+    let notes = client.do_list_notes(true).await;
+
+    let mut tx_spend: u64 = 0;
+    let mut tx_change: u64 = 0;
+    if let JsonValue::Array(spent_utxos) = &notes["spent_utxos"] {
+        for utxo in spent_utxos {
+            if utxo["spent"] == txid || utxo["unconfirmed_spent"] == txid {
+                tx_spend += utxo["value"].as_u64().unwrap_or(0);
```
Comment (Oscar-Pepper) [inline zingo-testutils/src/lib.rs]: I think this should panic if error during unwrap

```suggestion
                tx_spend += utxo["value"].as_u64().unwrap();
```

File: zingo-testutils/src/lib.rs [hunk 6]
```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
...
+    if let JsonValue::Array(unspent_orchard_notes) = &notes["unspent_orchard_notes"] {
+        for note in unspent_orchard_notes {
+            if note["created_in_txid"] == txid {
+                tx_change += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+
+    tx_spend - tx_change
+}
```
Comment (Oscar-Pepper) [inline zingo-testutils/src/lib.rs:519]: this function seems to return the total value transferred from a wallet which equals the amount sent + fee. for the zip317 test framework we are concerned with the fee only. Maybe we can determine the fee by calling this function on the recipient aswell but I feel a more elegant solution is to check the actual note structs within the transactions in the wallet and determine all information from the sender alone.

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]

File: zingo-testutils/src/lib.rs (hunk 1 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs]: change made in [6e3cb8b]

File: zingo-testutils/src/lib.rs (hunk 3 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs]: change made in [6e3cb8b]

File: zingo-testutils/src/lib.rs (hunk 4 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs]: change made in [d58d247]

File: zingo-testutils/src/lib.rs (hunk 5 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs]: change made in [6e3cb8b]

File: zingo-testutils/src/lib.rs (hunk 6 above)
Comment (idky137) [reply zingo-testutils/src/lib.rs:519]: i dont think its possible to calculate fees solely from notes known by the sending wallet, this is the reason ive been using total transfer value and subtracting the amount sent in do_send, but happy to make changes if you can suggest the code that will do this.

**Review by fluidvanadium — APPROVED** [review]
tested and acknowledged. this is a good failsafe

**Review by Oscar-Pepper — APPROVED** [review]
as we found, implementing a better fee calculation is going to be a large task and I think this is a good test for verifying zip317 for now.

**Pull request merged**